### ISSUE TITLE
SPDK build doesn't pick up modifications in libvfio-user files

### ISSUE THREAD

**Issue by tmakatos** [spdk/spdk#1930]
Modifying a file in libvfio-user and then building SPDK results in binaries not containing the change. The only realiable way to make this work is to run `make clean && (cd libvfio-user && make realclean && make && make install) && make`.

## Expected Behavior

Modifications to libvfio-user files followed by `make` in the top-level SPDK directory should results in binaries (e.g. `nvmf_tgt`) containing the relevant modifications.

## Current Behavior

Modifications to libvfio-user files followed by `make` in the top-level SPDK directory _do not_ result in binaries (e.g. `nvmf_tgt`) containing the relevant modifications.

## Steps to Reproduce

1. Modify a libvfio-user file, e.g. add an `abort()` at the beginning of `vfu_create_ctx`.
2. Run `make`.
3. Run an nvmf/vfio-user target: the `abort()` is not called.
4. Now retest with `make clean && (cd libvfio-user && make realclean && make && make install) && make`. The `abort()` does get called.

## Context (Environment including OS version, SPDK version, etc.)

Using https://review.spdk.io/gerrit/c/spdk/spdk/+/7617, but I'd expect this to trigger with latest master.

**Comment by changpe1** [comment]
@tmakatos after changing files in libvfio-user, you can run `make clean` && `make` in the SPDK home directory, it will link to the latest libvfio-user.

But yes, if you changed the source files in libvfio-user, and just run `make`, this won't take effect, because currently the nvmf_tgt binary is using static link, it can't detect that change.

A good way maybe we can use the shared library.

**Comment by changpe1** [comment]
@tmakatos Do you know which link parameter I can use to detect this?

**Comment by tmakatos** [comment]
As discussed on Slack, the problem is that libvfio-user.a is not a dependency of nvmf_tgt. Using libvfio-user.so, as you suggest, would be better.

**Comment by changpe1** [comment]
After some investigation, it's better for SPDK to use static library right now, there are many applications have dependency on nvmf library, if switched to use shared library, they need to compile and install libvfio-user library while they may not actually require it.

SPDK CI always installs the dependency libraries in a temporary directory, we need to configure the LD_LIBRARY_PATH manually, this isn't convenient for us.

**Comment by changpe1** [comment]
Another option maybe that when `make` is executed, SPDK will always re-link the binaries, this will fix the issue as well.

**Comment by tmakatos** [comment]
> Another option maybe that when make is executed, SPDK will always re-link the binaries, this will fix the issue as well.

Then you need to add libvfio-user.a as a dependency of nvmf.

**Comment by jimharris** [comment]
> > Another option maybe that when make is executed, SPDK will always re-link the binaries, this will fix the issue as well.
> 
> Then you need to add libvfio-user.a as a dependency of nvmf.

Yes, we need to add some additional plumbing to mk/spdk.app.mk so we can specify additional libraries that should cause a re-link.  We already do this for DPDK libraries, but don't do it for libraries from any of our other submodules (not just vfio-user, but also isa-l and ocf).

**Comment by changpe1** [comment]
I tested it and it works for me.
